libs/wingc/src/jsify.rs

```diff
@@ -1403,69 +1435,74 @@ impl<'ast> Visit<'ast> for FieldReferenceVisitor<'ast> {
-					message: format!("Cannot capture reassignable field '{}'", curr.text),
-					span: Some(curr.expr.span.clone()),
-				});
+						return;
+					}
 
-				return;
-			}
+					// if this type is not capturable, bail out
+					if !variable.type_.is_capturable() {
+						self.diagnostics.push(Diagnostic {
+							message: format!(
+								"Cannot capture field '{curr}' with non-capturable type '{}'",
+								variable.type_
+							),
+							span: Some(curr.span.clone()),
+						});
 
-			// if this type is not capturable, bail out
-			if !variable.type_.is_capturable() {
-				self.diagnostics.push(Diagnostic {
-					message: format!(
-						"Cannot capture field '{}' with non-capturable type '{}'",
-						curr.text, variable.type_
-					),
-					span: Some(curr.expr.span.clone()),
-				});
+						return;
+					}
 
-				return;
-			}
+					// okay, so we have a non-reassignable, capturable type.
+					// one more special case is collections. we currently only support
+					// collections which do not include resources because we cannot
+					// qualify the capture.
+					if let Some(inner_type) = variable.type_.collection_item_type() {
+						if inner_type.as_resource().is_some() {
+							self.diagnostics.push(Diagnostic {
+								message: format!(
+									"Capturing collection of resources is not supported yet (type is '{}')",
+									variable.type_,
+								),
+								span: Some(curr.span.clone()),
+							});
 
-			// okay, so we have a non-reassignable, capturable type.
-			// one more special case is collections. we currently only support
-			// collections which do not include resources because we cannot
-			// qualify the capture.
-			if let Some(inner_type) = variable.type_.collection_item_type() {
-				if inner_type.as_resource().is_some() {
-					self.diagnostics.push(Diagnostic {
-						message: format!(
-							"Capturing collection of resources is not supported yet (type is '{}')",
-							variable.type_,
-						),
-						span: Some(curr.expr.span.clone()),
-					});
-
-					return;
-				}
-			}
+							return;
+						}
+					}
 
-			// accumulate "curr" into capture
-			capture.push(curr);
-			index = index + 1;
+					// accumulate "curr" into capture
+					capture.push(curr);
+					index = index + 1;
 
-			// if "curr" is a collection, break here because the following
-			// components are going to be simple identifiers (TODO: is this a bug in
-			// how we model the API of collections?)
-			if variable.type_.collection_item_type().is_some() {
-				break;
+					// if "curr" is a collection, break here because the following
+					// components are going to be simple identifiers (TODO: is this a bug in
+					// how we model the API of collections?)
+					if variable.type_.collection_item_type().is_some() {
+						break;
+					}
+				}
+				ComponentKind::ClassType(_) => {
+					capture.push(curr);
+					index = index + 1;
+				}
+				ComponentKind::Unsupported => panic!("all components should be valid at this point"),
```

Comment:
I think it's not possible because all collection types (which are the `Unsupported` components) are treated as if anything done on them is already inflight, so we break before we reach this. This is preserving the previous logic (which I think was an assert).